### PR TITLE
arm64/imx9: add imx93-evk ddr training

### PR DESCRIPTION
**Pull request by eenurkka**
This performs the DDR training for imx93-evk. In addition to the source code, it downloads binaries which are included in the final image. The bootloader must be ARCH_CORTEX_A53 instead of A55 due to atomic instructions that don't work with the OCRAM / EL3 combination.

## Summary

This add DDR training so that the system may use the external DDR memory.

## Impact

All imx93 bootloaders will now also initialize the DDR memory present.

## Testing

imx93-evk with SD-card boot (bootloader on SD-card), with NSH used for simple manual tests; mw command enabled for direct memory access.

### CODE REVIEW

File: arch/arm64/src/imx9/ddr/hardware/imx9_ddr_training.h [hunk 1]
```diff
@@ -0,0 +1,946 @@
...
+ ****************************************************************************/
+
+#ifndef __ARCH_ARM64_SRC_IMX9_DDR_HARDWARE_IMX9_DDR_TRAINING_H
+#define __ARCH_ARM64_SRC_IMX9_DDR_HARDWARE_IMX9_DDR_TRAINING_H
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#define FILE_DMEM_1D "ddr/firmware-imx-8.23/firmware/ddr/synopsys/lpddr4_dmem_1d_v202201.bin"
```
Comment (pussuw) [inline arch/arm64/src/imx9/ddr/hardware/imx9_ddr_training.h:28]: Move these elsewhere; they don't need to be visible in the HW description. Should be board specific stuff ?

File: arch/arm64/src/imx9/Make.defs [hunk 2]
```diff
@@ -78,8 +82,11 @@ ifeq ($(CONFIG_IMX9_FLEXSPI_NOR), y)
 endif
 
 ifeq ($(CONFIG_IMX9_BOOTLOADER), y)
- CHIP_CSRCS += imx9_system_ctl.c
- CHIP_CSRCS += imx9_trdc.c
+  CHIP_CSRCS += imx9_system_ctl.c
+  CHIP_CSRCS += imx9_trdc.c
+  ifeq ($(CONFIG_IMX9_DDR_TRAINING),y)
```
[the remark this answers is not in the document]
Comment (eenurkka) [reply arch/arm64/src/imx9/Make.defs]: I recall in Kconfig, they kinda depend, but separated these

[the remark this answers is not in the document]
Comment (pussuw) [reply arch/arm64/src/imx9/Make.defs]: Yes in practice they do, I guess

**Review by acassis — APPROVED** [review]

File: arch/arm64/src/imx9/ddr/hardware/imx9_ddr_training.h [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#ifndef ARRAY_SIZE
+#define ARRAY_SIZE(x) (sizeof(x) / sizeof((x)[0]))
+#endif
+
+#define STR2(m) #m
+#define STR(m) STR2(m)
```
[the remark this answers is not in the document]
Comment (pussuw) [reply arch/arm64/src/imx9/ddr/hardware/imx9_ddr_training.h:33]: STRINGIFY() in the same header does same as STR() here.

File: arch/arm64/src/imx9/ddr/hardware/imx9_ddr_training.h [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+#ifndef __ARCH_ARM64_SRC_IMX9_DDR_HARDWARE_IMX9_DDR_TRAINING_H
+#define __ARCH_ARM64_SRC_IMX9_DDR_HARDWARE_IMX9_DDR_TRAINING_H
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#ifndef ARRAY_SIZE
+#define ARRAY_SIZE(x) (sizeof(x) / sizeof((x)[0]))
```
[the remark this answers is not in the document]
Comment (eenurkka) [reply arch/arm64/src/imx9/ddr/hardware/imx9_ddr_training.h:29]: Right!

File: arch/arm64/src/imx9/ddr/hardware/imx9_ddr_training.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (eenurkka) [reply arch/arm64/src/imx9/ddr/hardware/imx9_ddr_training.h:33]: Yes, switched to use STRINGIFY

File: arch/arm64/src/imx9/ddr/hardware/imx9_ddr_training.h [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+ ****************************************************************************/
+
+#ifndef __ARCH_ARM64_SRC_IMX9_DDR_HARDWARE_IMX9_DDR_TRAINING_H
+#define __ARCH_ARM64_SRC_IMX9_DDR_HARDWARE_IMX9_DDR_TRAINING_H
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#define MHZ(x) ((x) * 1000000UL)
```
Comment (pussuw) [inline arch/arm64/src/imx9/ddr/hardware/imx9_ddr_training.h:28]: I don't think MHZ() is used anywhere, please remove it

Comment (eenurkka) [reply arch/arm64/src/imx9/ddr/hardware/imx9_ddr_training.h]: Good catch! fixed.

**Review by pussuw — APPROVED** [review]
LGTM!

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**